src/main/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/ReadsPreprocessingPipeline.java

```diff
@@ -0,0 +1,173 @@
+            shortName = StandardArgumentDefinitions.INPUT_SHORT_NAME, fullName = StandardArgumentDefinitions.INPUT_LONG_NAME,
+            optional = false)
+    protected String bam;
+
+    @Argument(doc = "", shortName = "O", fullName = "output", optional = false)
+    protected String output;
+
+    @Argument(doc = "", shortName = "R", fullName = "reference", optional = false)
+    protected String referenceName;
+
+    @Argument(doc = "", shortName = "BQSRKnownVariants", fullName = "baseRecalibrationKnownVariants", optional = true)
+    protected List<String> baseRecalibrationKnownVariants;
+
+    @ArgumentCollection
+    protected IntervalArgumentCollection intervalArgumentCollection = new OptionalIntervalArgumentCollection();
+
+    @Override
+    protected void setupPipeline( Pipeline pipeline ) {
+        // Load the reads.
+        final ReadsDataflowSource readsDataflowSource = new ReadsDataflowSource(bam, pipeline);
+        final SAMFileHeader readsHeader = readsDataflowSource.getHeader();
+        final List<SimpleInterval> intervals = intervalArgumentCollection.intervalsSpecified() ? intervalArgumentCollection.getIntervals(readsHeader.getSequenceDictionary()):
+                                                                                                 getAllIntervalsForReference(readsHeader.getSequenceDictionary());
+
+        final PCollectionView<SAMFileHeader> headerSingleton = pipeline.apply(Create.of(readsHeader)).setCoder(SerializableCoder.of(SAMFileHeader.class)).apply(View.<SAMFileHeader>asSingleton());
+        final PCollection<GATKRead> initialReads = readsDataflowSource.getReadPCollection(intervals);
+
+        // Apply MarkDuplicates to produce updated GATKReads.
+        final PCollection<GATKRead> markedReads = initialReads.apply(new MarkDuplicatesStub(headerSingleton));
+
+        // Load the Variants and the Reference and join them to reads.
+        final VariantsDataflowSource variantsDataflowSource = new VariantsDataflowSource(baseRecalibrationKnownVariants, pipeline);
+
+        RefAPISource refAPISource = new RefAPISource();
+        Map<String, String> referenceNameToIdTable = RefAPISource.buildReferenceNameToIdTable(pipeline.getOptions(), referenceName);
+        RefAPIMetadata refAPIMetadata = new RefAPIMetadata(referenceName, referenceNameToIdTable);
+
+        final PCollection<KV<GATKRead, ReadContextData>> readsWithContext = AddContextDataToRead.add(markedReads, refAPISource, refAPIMetadata, variantsDataflowSource);
+
+        // Apply BQSR.
+        final PCollection<RecalibrationTables> recalibrationReports = readsWithContext.apply(new BaseRecalibratorStub(headerSingleton));
+        final PCollectionView<RecalibrationTables> mergedRecalibrationReport = recalibrationReports.apply(View.<RecalibrationTables>asSingleton());
+
+        final PCollection<GATKRead> finalReads = markedReads.apply(new ApplyBQSRStub(headerSingleton, mergedRecalibrationReport));
```

Comment:
ApplyBQSR doesn't take the RecalibrationTables as input, but that's OK I can change that later when I merge it in.